File: 数仓表结构/asset_package/src/main/java/com/weshare/interfaces/impl/BagInfoImpl.java
package com.weshare.interfaces.impl;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.weshare.entity.BagInfo;
import com.weshare.interfaces.AssetFiles;
import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.SparkSession;

import java.io.Serializable;

/**
 * Created by mouzwang on 2020-11-30 15:50
 */
public class BagInfoImpl implements AssetFiles, Serializable {
    private static final long serialVersionUID = 1L;

    public Dataset<BagInfo> process(Dataset<String> dataset) {
        Dataset<BagInfo> mapedDS = dataset.map(new MapFunction<String, BagInfo>() {
            @Override
            public BagInfo call(String value) throws Exception {
                BagInfo bagInfo = new BagInfo();
                JsonObject jsonObject = new JsonParser().parse(value).getAsJsonObject();
                bagInfo.setProjectId(jsonObject.get("project_id").getAsString());
                bagInfo.setBagId(jsonObject.get("bag_id").getAsString());
                bagInfo.setBagName(jsonObject.get("bag_name").getAsString());
                bagInfo.setBagStatus(jsonObject.get("bag_status").getAsString());
                bagInfo.setBagRemainPrincipal(jsonObject.get("bag_remain_principal").getAsBigDecimal());
                bagInfo.setBagDate(jsonObject.get("bag_date").getAsString());
                bagInfo.setInsertDate(jsonObject.get("insert_date").getAsString());
                return bagInfo;
            }
        }, Encoders.bean(BagInfo.class));
        return mapedDS;
    }


    @Override
    public void insertData(Dataset<String> dataset, SparkSession sparkSession, String fileId) {
        Dataset<BagInfo> processDS = process(dataset);
        processDS.registerTempTable("temp_bag_info");
        sparkSession.sql("insert into table dim.bag_info partition (bag_id = '" + fileId + "') " +
                "select projectId,bagName,bagStatus,bagRemainPrincipal,bagDate,insertDate from " +
                "temp_bag_info ");
    }

    @Override
    public void updateData(Dataset<String> dataset, SparkSession sparkSession, String fileId) {
        Dataset<BagInfo> processDS = process(dataset);
        processDS.registerTempTable("temp_bag_info");
        sparkSession.sql("insert overwrite table dim.bag_info partition (bag_id = '" + fileId + "') " +
                "select projectId,bagName,bagStatus,bagRemainPrincipal,bagDate,insertDate from " +
                "temp_bag_info ");
    }

    @Override
    public void deleteData(Dataset<String> dataset, SparkSession sparkSession, String fileId) {
        sparkSession.sql("alter table dim.bag_info drop partition (bag_id = '" + fileId + "')");
        sparkSession.sql("alter table dim.bag_due_bill_no drop partition (bag_id = '" + fileId + "')");
    }

}

File: 数仓表结构/asset_package/src/main/scala/com/weshare/interface/impl/BagInfoImpl.scala
package com.weshare.interface.impl

import com.weshare.interface.AssetFile
import org.apache.spark.sql.functions.lit
import org.apache.spark.sql.{DataFrame, Dataset, SparkSession}

/**
  * @author ximing.wei 2021-05-27 21:21:46
  */
class BagInfoImpl extends AssetFile {
    override def insertData(): Unit = {
        logger.info(logBase, "执行 插入 语句！")

        logger.info(logBase, "将获取的数据，按字段创建临时表  bag_info_tmp ！")
        data.select(
            lit("project_id"),
            lit("bag_id"),
            lit("bag_name"),
            lit("bag_status"),
            lit("bag_remain_principal"),
            lit("bag_date"),
            lit("insert_date")
        ).createOrReplaceTempView("bag_info_tmp")

        logger.info(logBase, "执行 插入 语句 HQL！")
        sparkSession.sql(
            """
              |insert overwrite table dim.bag_info partition(bag_id)
              |select
              |  project_id,
              |  bag_name,
              |  bag_status,
              |  bag_remain_principal,
              |  bag_date,
              |  insert_date,
              |  bag_id
              |from bag_info_tmp
            """.stripMargin)
        logger.info(logBase, "插入 语句执行 完成！")
    }

    override def updateData(): Unit = {
        logger.info(logBase, "执行 更新 语句！")

        logger.info(logBase, "调用 插入 语句！")
        insertData()
        logger.info(logBase, "更新 语句执行 完成！")
    }

    override def deleteData(): Unit = {
        logger.info(logBase, "执行 删除 语句！")

        logger.info(logBase, "开始执行 删除 语句！")
        data.select(lit("bag_id")).collect().foreach(
            row => {
                val fieldsMap = row.getValuesMap[Any](List("bag_id"))
                val bagId = fieldsMap("bag_id").toString

                logger.info(logBase, s"删除 bag_info 的包编号：$bagId！")
                sparkSession.sql(s"ALTER TABLE dim.bag_info DROP IF EXISTS PARTITION (import_id = '$bagId')")

                logger.info(logBase, s"删除 bag_due_bill_no 的包编号：$bagId！")
                sparkSession.sql(s"ALTER TABLE dim.bag_due_bill_no DROP IF EXISTS PARTITION (import_id = '$bagId')")
            }
        )
        logger.info(logBase, "删除 语句执行 完成！")
    }
}
